jepsen/src/jepsen/nemesis/combined.clj

```diff
@@ -249,6 +263,133 @@
                              :fs    #{:strobe-clock}
                              :color "#A0E9E3"}}})))
 
+(defn wan-nemesis
+  "A nemesis which simulates running the cluster on a wide-area network, e.g.
+  a cluster of nodes split across multiple different AWS regions, i.e. higher
+  average latency, larger latency variance and spottier connectivity.
+
+  This nemesis responds to the following f's:
+
+      :drop     (Drop traffic between two nodes.)
+      :slow     (Delays network packets.)
+      :flaky    (Introduces randomzied packet loss.)
+      :heal     (End all traffic drops and restore network to fast operation.)
+      :fast     (Removes packet loss and delays.)"
+  [_opts]
+  (reify
+    n/Reflection
+    (fs [_this] #{:drop :slow :flaky :heal :fast})
+
+    n/Nemesis
+    (setup! [this test]
+      (net/heal! (:net test) test)
+      (net/fast! (:net test) test)
+      this)
+
+    (invoke! [this test op]
+      (let [res (case (:f op)
+                  :drop  (net/drop! (:net test) test (-> op :value :src) (-> op :value :dest))
+                  :slow  (net/slow! (:net test) (assoc test :nodes (-> op :value :targets))
+                                    {:mean         (-> op :value :mean)
+                                     :variance     (-> op :value :variance)
+                                     :distribution (-> op :value :distribution)})
+                  :flaky (net/flaky! (:net test) (assoc test :nodes (-> op :value :targets)))
+                  :heal  (net/heal!  (:net test) test)
+                  :fast  (net/fast!  (:net test) test))]
+        (assoc op :value res)))
+
+    (teardown! [_this _test])))
+
+(defn wan-generators
+  "A map with a :generator and :final-generator for WAN-related operations
+  Options are from nemesis-package."
+  [opts]
+  (let [db      (:db opts)
+        faults  (:faults opts)
+        drop?   (contains? faults :drop)
+        slow?   (contains? faults :slow)
+        flaky?  (contains? faults :flaky)
```

Comment:
So... we've got a bit of a problem here, in that all of these could conceivably apply *either* to individual nodes/pairs of nodes, *or* to datacenters. I'm inclined to hold off on merging this into nemesis.combined until we can figure out how to get those things encoded in a way that makes sense. Maybe datacenters should be a protocol that databases support, and we add datacenter-specific *targets* to the existing partition nemesis?